Cython/Compiler/Dataclass.py

```diff
@@ -0,0 +1,611 @@
+# functions to transform a c class into a dataclass
+
+from collections import OrderedDict
+
+from .Errors import error, warning
+from . import ExprNodes, Nodes, PyrexTypes
+from .Code import UtilityCode
+from .Visitor import VisitorTransform
+from . import UtilNodes, Builtin, Naming
+from .StringEncoding import BytesLiteral, EncodedString
+from .TreeFragment import TreeFragment
+from .ParseTreeTransforms import (NormalizeTree, SkipDeclarations, AnalyseDeclarationsTransform,
+                                  MarkClosureVisitor)
+
+def _make_module_callnode(pos, name):
+    loader_utilitycode = UtilityCode.load_cached("%sModuleLoader" % name, "Dataclasses.c")
+    return ExprNodes.PythonCapiCallNode(pos, "__Pyx_Load%sModule" % name,
+                                PyrexTypes.CFuncType(PyrexTypes.py_object_type, []),
+                                utility_code = loader_utilitycode,
+                                args=[])
+
+def make_dataclass_module_callnode(pos):
+    return _make_module_callnode(pos, "Dataclass")
+def make_typing_module_callnode(pos):
+    return _make_module_callnode(pos, "Typing")
+
+_INTERNAL_DEFAULTSHOLDER_NAME = EncodedString('__pyx_dataclass_defaults')
+
+def make_common_utilitycode(scope):
+    scope.global_scope().use_utility_code(
+        UtilityCode.load_cached("DataclassModuleLoader", "Dataclasses.c"))
+
+
+class RemoveAssignments(VisitorTransform, SkipDeclarations):
+    def __init__(self, names):
+        super(RemoveAssignments, self).__init__()
+        self.names = names
+        self.removed_assignments = {}
+
+    def visit_CClassNode(self, node):
+        self.visitchildren(node)
+        return node
+
+    def visit_PyClassNode(self, node):
+        return node  # go no further
+
+    def visit_FuncDefNode(self, node):
+        return node  # go no further
+
+    def visit_SingleAssignmentNode(self, node):
+        if node.lhs.is_name and node.lhs.name in self.names:
+            if node.lhs.name in self.removed_assignments:
+                warning(node.pos, ("Multiple assignments for '%s' in dataclass; "
+                                   "using most recent") % node.lhs.name, 1)
+            self.removed_assignments[node.lhs.name] = node.rhs
+            return []
+        return node
+
+    # I believe cascaded assignment is always a syntax error with decorators
+    # so can be ignored
+
+    def visit_Node(self, node):
+        self.visitchildren(node)
+        return node
+
+class _MISSING_TYPE:
+    pass
+MISSING = _MISSING_TYPE()
+
+def process_class_get_fields(node):
+    _TrueNode = ExprNodes.BoolNode(node.pos, value=True)
+    _FalseNode = ExprNodes.BoolNode(node.pos, value=False)
+    _NoneNode = ExprNodes.NoneNode(node.pos)
+
+    class Field:
+        # store fields as nodes so they can be used in code construction
+        # more readily; define class inside function so default arguments
+        # have a useful pos
```

Comment:
But public fields already have property nodes, right? Isn't that enough to handle them?

Reply:
The CPython dataclass implementation defines [`field`](https://docs.python.org/3/library/dataclasses.html#dataclasses.field) which is used to give extra control - for example for turning saying "this attribute doesn't appear in `__repr__`" or for setting up a default factory function.

`field` ends up being used in two places: controlling the code generation and as inspectable information in `__dataclass_fields__`. In the standard library module the same structure is naturally used for both. In Cython that doesn't quite work as naturally. Even so I found it useful to replicate the `dataclasses.field` structure at the code generation stage to keep track of what options are set for each attribute.

When I was writing this I was copying from the standard library module quite a bit so it made sense to try to use similar structures.

I'll try to get some of this in comments next to the code though since the comments currently next to the code aren't particularly helpful.